package.json

```diff
@@ -77,8 +76,10 @@
     "prettier": "3.2.5",
     "rimraf": "^5.0.7",
     "rollup": "^4.13.0",
+    "rollup-plugin-esbuild": "^6.1.1",
     "semver": "^7.6.2",
     "simple-git-hooks": "^2.11.1",
+    "terser": "^5.31.0",
```

Comment:
This are needing because when running `vite` in `playground/lagacy` of `rollup` in `packages/vite`, pnpm use the top-level `vite`/`rollup` and then don't see `terser`/`rollup-plugin-esbuild`.

I know have a good example of why yarn v1 is still less painful for me

Reply:
I cannot reproduce the case for `terser`, do you have an example for that? Re `rollup-plugin-esbuild`, seems to be an issue only for `--configPlugin esbuild`. Not sure why Rollup is executed in the wrong place.

Reply:
It works for me.

Reply:
After a `pnpm dev` in `vite`? I look into this a bit more tomorrow because I don't really understand what is happening.

Reply:
This is really strange, after this change, when running vite dev inside `playground/legacy`, the value of `config.root` goes from `playground/legacy` to `packages/plugin-legacy/dist`

Reply:
I tried `pnpm dev` and `pnpm build` in Vite, and both worked for me in the legacy playground. I also don't see the root changing. Maybe you can try nuking node_modules and re-installing?

Reply:
I push a version without terser at the root and re installed node_modules from scratch.
Can you confirm that with the latest version of the branch, running `pnpm i && pnpm dev` and then in another terminal session: `cd playground/legacy && pnpm build` you don't get any error?

Reply:
Ah ok now I'm getting it. Only happens with `pnpm dev` in Vite. Weird 🤔

Reply:
I found it, it's coming from here: https://github.com/vitejs/vite/blob/2dc327833d98b578303e3743c4a543f540a3c7b5/packages/plugin-legacy/src/index.ts#L780-L781

Installing terser in the `plugin-legacy` package fixes it.

Reply:
But that's so weird it doesn't happen on main, I don't understand

Reply:
With some digging, it seems like terser is being loaded twice. Once from `playground/legacy`, second from `plugin-legacy/dist`. In main and this PR's `build`, it worked because this code caches the `terserPath` on the first attempt (`playground/legacy`):

https://github.com/vitejs/vite/blob/2dc327833d98b578303e3743c4a543f540a3c7b5/packages/vite/src/node/plugins/terser.ts#L17-L18

It didn't work in `pnpm dev` because the build had that code duplicated in separate chunks, so the variable needs to be re-computed on the second attempt, leading to the issue. I wonder if we could combine the chunks or how it happened. (Search for `terserPath` in `dist/node/cli.js` and `dist/node/index.js`)

Reply:
Oh yes I could try to enable `codeSplit: true` for the node bundle in dev.

From my initial description:
> but without chunking. It generates two index.js & cli.js of ~600kbs which is fine

Reply:
I just try adding code splitting and esbuild is really less good than rollup at this, instead of 3 nice endtrypoints and one shared chunks, I get a chunks folder with
```
build-BHFHFIPJ.js
chunk-2JJU2EFR.js
chunk-4BPKRV5Y.js
optimizer-6WZRCE2Z.js
preview-QGORIEUU.js
server-FT3KPPC2.js
```

and the code from the constant file is not in chunks, which makes this code fails:
```js
const { version } = JSON.parse(
  readFileSync(new URL('../../package.json', import.meta.url)).toString(),
)
// [...]
export const VITE_PACKAGE_DIR = resolve(
  // import.meta.url is `dist/node/constants.js` after bundle
  fileURLToPath(import.meta.url),
  '../../..',
)
``` 

Adding terser to the plugin legacy seems simpler, but at the same time not having shared chunk between cli & index is maybe more an issue I though at the begging and subtitle bug could occur.

Reply:
Yeah I feel like there could be subtle bugs if we don't share a chunk. Maybe with some refactoring we could keep the constant work in both dev and build, but if not, perhaps we can go with a middle ground for now and merge in the build changes only? The speedup should still help.

Reply:
I'll try making the constant chunks works with esbuild, otherwise I'll all the other improvements like using esbuild + no type generation that should still give good benefits in dev too

Reply:
updated and avoided the `../../..` reference issue by not putting esbuild chunks into a sub-folder. This `___require` hack is still there but the instant watch mode with source maps is worth it!

Reply:
Awesome! We may need to be careful of how esbuild chunks this in the future, but I think it's worth the tradeoff for now.